x/cheqd/types/did_validation.go

```diff
@@ -0,0 +1,220 @@
+//	if !strings.Contains(didUrl, "#") {
+//		return false
+//	}
+//
+//	did, _ := SplitDidUrlIntoDidAndFragment(didUrl)
+//	return IsValidDid(prefix, did)
+//}
+//
+//func ResolveId(did string, methodId string) string {
+//	result := methodId
+//
+//	methodDid, methodFragment := SplitDidUrlIntoDidAndFragment(methodId)
+//	if len(methodDid) == 0 {
+//		result = did + "#" + methodFragment
+//	}
+//
+//	return result
+//}
+//
+//func IsNotValidDIDArray(prefix string, array []string) (bool, int) {
+//	for i, did := range array {
+//		if !IsValidDid(prefix, did) {
+//			return true, i
+//		}
+//	}
+//
+//	return false, 0
+//}
+//
+//func IsNotValidDIDArrayFragment(prefix string, array []string) (bool, int) {
+//	for i, did := range array {
+//		if !IsDidFragment(prefix, did) {
+//			return true, i
+//		}
+//	}
+//
+//	return false, 0
+//}
+//
+//func IsValidDid(prefix string, did string) bool {
+//	if len(did) == 0 {
+//		return false
+//	}
+//
+//	if !DidForbiddenSymbolsRegexp.MatchString(did) {
+//		return false
+//	}
+//
+//	// FIXME: Empty namespace must be allowed even if namespace is set in state
+//	// https://github.com/cheqd/cheqd-node/blob/main/architecture/adr-list/adr-002-cheqd-did-method.md#method-specific-identifier
+//	return strings.HasPrefix(did, prefix)
+//}
+
+
+// DID
+
+func ValidateDID(did string, method string, allowedNamespaces []string) error {
+	method, namespace, unique_id := SplitDID(did)
+
+	// check method
+	if method != method {
+		return ErrStaticDIDBadMethod.Wrap(method)
+	}
+	// check namespaces
+	if !DidNamespaceRegexp.MatchString(namespace) || !utils.Contains(allowedNamespaces, namespace) {
+		return ErrStaticDIDNamespaceNotAllowed.Wrap(namespace)
+	}
+	// check unique-id
+	err := ValidateUniqueId(unique_id)
+
+	return err
+}
+
+func ValidateUniqueId(unique_id string) error {
+	// Length should be 16 or 32 symbols
+	if len(unique_id) != 16 && len(unique_id) != 32 {
+		return ErrStaticDIDBadUniqueIDLen.Wrap(unique_id)
+	}
+	// Base58 check
+	if !UniqueIDRegexp.MatchString(unique_id) {
+		return ErrStaticDIDNotBase58ID.Wrap(unique_id)
+	}
+
+	return nil
+}
+
+func IsValidDID(did string, method string, allowedNamespaces []string) bool {
+	err := ValidateDID(did, method, allowedNamespaces)
+	return err == nil
+}
+
+// SplitDID panics if did is not valid
+func SplitDID(did string) (method string, namespace string, id string) {
+	// Example: did:cheqd:testnet:base58str1ng1111
+	// match [0] - the whole string
+	// match [1] - cheqd                - method
+	// match [2] - :testnet
+	// match [3] - testnet              - namespace
+	// match [4] - base58str1ng1111     - id
+	match := SplitDIDRegexp.FindStringSubmatch(did)
+	if len(match) > 0 {
+		return match[1], match[3], match[4]
+	}
+
+	return "", "", ""
+}
+
+// SplitDIDUrl panics if did cannot be splitted properly
+func SplitDIDUrl(didUrl string) (did string, path string , query string, fragment string) {
+	match := SplitDIDURL.FindStringSubmatch(didUrl)
```

Comment:
Let's check for matches count

Reply:
Shouldn't we check for matches count?

Reply:
It matches only once for Golang regexps... I checked it

Reply:
BTW, The last changes in regexp doesn't match in case of fragment being `#key123???#####`